### ISSUE TITLE
PRNGKey input issues

### ISSUE THREAD

**Issue by jakevdp** [google/jax#4745]
I've found a few issues with `jax.random.PRNGKey` for corner-cases of input.

First, Python integers with values between the maximum representable int64 and the maximum representable uint64 lead to an error, while dtyped values in the same range do not:
```python
import jax.numpy as jnp
from jax import config, random
config.update('jax_enable_x64', True)

print(random.PRNGKey(jnp.uint64(1 << 63)))
# [2147483648          0]
print(random.PRNGKey(1 << 63))
# TypeError: ufunc 'right_shift' not supported for the input types, and the inputs could not be safely 
# coerced to any supported types according to the casting rule ''safe''
```
I believe the correct fix for this is to change the handling of Python ints so that the results are consistent with their dtyped counterparts.

Secondly, the key that results from passing a negative seed changes depending on whether that seed is a Python integer or a dtyped integer:
```python
print(random.PRNGKey(-1000))
# [4294967295 4294966296]
print(random.PRNGKey(jnp.int32(-1000)))
# [         0 4294966296]
```
The best fix here is less obvious: we could fix input handling so that the two give consistent results, but this would silently change some PRNG outputs in the next JAX release.

Better might be to explicitly raise a ValueError for negative seeds, which would would be a potential breakage for some users, but would at least be a loud & obvious one.

Third, there is a breakage of JIT invariance when passing a large Python or numpy scalar in X32 mode:
```python
from jax import random, jit
print(random.PRNGKey(1 << 50))
# [262144      0]
print(jit(random.PRNGKey)(1 << 50))
# [0 0]
```
This one is tricky: the only way to make this invariant to JIT is to truncate Python & numpy inputs to 32-bit when working in x32 mode. But the effect of this would be that identical large seeds would lead to entirely different random sequences when flipping the x64 flag. I think JIT invariance here is probably more important than x64 invariance, however.

Thoughts?

**Comment by mattjj** [comment]
Nice finds! To me all these solutions sound reasonable. On the second one, maybe I have a slight lean towards loud error, but both seem defensible.

**Comment by jakevdp** [comment]
I'm leaning slightly toward allowing negative numbers, because when converted to a uint (which is effectively what the Python int version does) it provides easier access to filling most/all bits.

**Comment by jakevdp** [comment]
I found another potential issue: our special handling of int and numpy types outside x64 mode leads to breaking of jit invariance in x32 mode:
```python
from jax import random, jit
print(random.PRNGKey(1 << 50))
# [262144      0]
print(jit(random.PRNGKey)(1 << 50))
# [0 0]
```
I think the only way we can achieve consistency across x32/x64 as well as jit/non-jit would be to cast all 64-bit inputs to 32-bit. The problem with that is that fully half of the key space would be inaccessible to `PRNGKey()`.

**Comment by jakevdp** [comment]
JIT invariance addressed in #4877. The differences between equal seeds of differing dtype seem less important, so I avoided that to limit potential breakages of existing code.
